### PR TITLE
UCT/IB: preparation for DC transport

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe please take a look

### CODE REVIEW

File: src/uct/ib/base/ib_iface.h [hunk 1]
```diff
@@ -256,4 +256,23 @@ ucs_status_t uct_ib_iface_arm_tx_cq(uct_ib_iface_t *iface);
 
 ucs_status_t uct_ib_iface_arm_rx_cq(uct_ib_iface_t *iface, int solicited);
 
+#define UCT_IB_IFACE_VERBS_FOREACH_RXWQE(iface, i, hdr, wc, wc_count) \
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.h]: marco parameters should start with _

File: src/uct/ib/base/ib_iface.h [hunk 2]
```diff
@@ -256,4 +256,23 @@ ucs_status_t uct_ib_iface_arm_tx_cq(uct_ib_iface_t *iface);
...
+                                                      (uct_ib_iface_recv_desc_t *)(uintptr_t)wc[i].wr_id); \
+        VALGRIND_MAKE_MEM_DEFINED(hdr, wc[i].byte_len); \
+        UCS_INSTRUMENT_RECORD(UCS_INSTRUMENT_TYPE_IB_RX, __FUNCTION__, \
+                              wc[i].wr_id, wc[i].status); \
+               1; }); ++i) 
+
+#define UCT_IB_IFACE_VERBS_FOREACH_TXWQE(iface, i, wc, wc_count) \
+    for (i = 0; i < wc_count && ({ \
+        if (ucs_unlikely(wc[i].status != IBV_WC_SUCCESS)) { \
+            ucs_fatal("iface=%p: send completion %d with error: %s wqe: %p wr_id: %llu", iface, i, ibv_wc_status_str(wc[i].status), &wc[i], (unsigned long long)wc[i].wr_id); \
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.h]: pls split line

File: src/uct/ib/rc/accel/rc_mlx5_ep.c [hunk 3]
```diff
@@ -351,11 +344,13 @@ uct_rc_mlx5_ep_bcopy_post(uct_rc_mlx5_ep_t *ep, unsigned opcode, unsigned length
                           /* RDMA */ uint64_t rdma_raddr, uct_rkey_t rdma_rkey,
                           int force_sig, uct_rc_iface_send_desc_t *desc)
 {
-    desc->super.sn = ep->tx.wq.sw_pi;
+    uint16_t sn;
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_ep.c]: it's more efficient to remember sn inside desc immediately than to use 'sn' variable

File: src/uct/ib/rc/base/rc_iface.c [hunk 4]
```diff
@@ -393,26 +393,36 @@ static UCS_CLASS_CLEANUP_FUNC(uct_rc_iface_t)
...
     struct ibv_qp *qp;
     int inline_recv = 0;
 
+    IB_RC_QP_TYPE_CHECK(qp_type);
     memset(&qp_init_attr, 0, sizeof(qp_init_attr));
     qp_init_attr.qp_context          = NULL;
     qp_init_attr.send_cq             = iface->super.send_cq;
     qp_init_attr.recv_cq             = iface->super.recv_cq;
-    qp_init_attr.srq                 = iface->rx.srq;
+    if (qp_type == IBV_QPT_RC) {
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.c]: why is that needed?

File: src/uct/ib/rc/base/rc_iface.h [hunk 5]
```diff
@@ -23,11 +23,60 @@
 #define UCT_RC_MAX_ATOMIC_SIZE      sizeof(uint64_t)
 #define UCR_RC_QP_MAX_RETRY_COUNT   7
 
+#define UCT_RC_CHECK_AM_SHORT(am_id, length, max_inline) \
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.h]: leading _ for macro args

File: test/gtest/uct/uct_test.h [hunk 6]
```diff
@@ -183,6 +183,7 @@ std::ostream& operator<<(std::ostream& os, const resource* resource);
 #define UCT_TEST_IB_TLS \
     rc_mlx5, \
     rc, \
+    dc, \
```
Comment (yosefe) [inline test/gtest/uct/uct_test.h]: probably not needed?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1347/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1972/ for details.

File: src/uct/ib/rc/base/rc_iface.c (hunk 4 above)
Comment (alex-mikheev) [inline src/uct/ib/rc/base/rc_iface.c]: it is needed to create dc qp

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1353/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1978/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1354/ for details (Mellanox internal link).
